src/cluster.js

```diff
@@ -1,77 +1,109 @@
-      this._ecs.runTask(params, (err, data) => {
-        if (err) {
-          reject(err)
-        } else if (data.tasks.length == 0) {
-          reject(new TaskNotPresentInECSCallbackError())
-        } else {
-          resolve(data.tasks[0])
-        }
-      })
-    })
+  start() {
+    this._server.start()
+    this._monitoringCluster = true
+    this._monitorCluster()
   }
 
-  _ecsClustName() {
-    return process.env.ECS_CLUSTER || "default"
+  startBuild(build) {
+    let container = this._firstAvailableContainer()
+
+    if (container) {
+      return this._startBuildOnContainer(build, container)
+    } else {
+      return Promise.reject(new NoContainersAvailableError())
+    }
   }
 
-  _ecsDescriptionParams() {
-    return {
-      clusters: [ this._ecsClustName() ]
+  stop() {
+    this._server.stop()
+    this._monitoringCluster = false
+  }
+
+  stopBuild(buildID) {
+    const container = this._findBuildContainer(buildID)
+    if (container) {
+      clearTimeout(container.timeout)
+      container.build = undefined
     }
   }
 
-  _ecsTaskDefinition() {
-    return process.env.ECS_TASK
+  _firstAvailableContainer() {
+    return this._containers.find(container => !container.build)
+  }
+
+  _findBuildContainer(buildID) {
+    return this._containers.find(container => {
+      return container.build.buildID === buildID
+    })
   }
 
-  _ecsTaskParams(containerOverrides) {
-    return {
-      taskDefinition: this._ecsTaskDefinition(),
-      cluster: this._ecsClustName(),
-      overrides: { containerOverrides: [ containerOverrides ] }
+  _findContainer(guid) {
+    return this._containers.find(container => container.guid === guid)
+  }
+
+  _monitorCluster() {
+    if (this._monitoringCluster) {
+      this._apiClient.fetchBuildContainers().then(containers => {
+        this._resolveNewContainers(containers)
+      }).then(() => {
+        setTimeout(() => {
+          this._monitorCluster()
+        }, 60 * 1000)
+      })
     }
   }
-}
 
-class ClusterNotPresentInECSCallbackError extends Error {
-  constructor() {
-    super("ECS.DescribeClusters called back with a response that did not include any clusters")
+  _resolveNewContainers(containers) {
+    this._containers = containers.map(newContainer => {
+      const existingContainer = this._findContainer(newContainer.guid)
+
+      if (existingContainer) {
+        return Object.assign(newContainer, {
+          build: existingContainer.build,
+          timeout: timeout,
+        })
+      } else {
+        return newContainer
+      }
+    })
+  }
+
+  _startBuildOnContainer(build, container) {
+    container.build = build
+    container.timeout = setTimeout(() => {
+      this.stopBuild(build.buildID)
+    }, 300 * 1000)
+    return this._apiClient.updateBuildContainer(
+      container,
+      build.containerEnvironment
+    ).catch(error => {
+      container.build = undefined
+      clearTimeout(container.timeout)
+      throw error
```

Comment:
Good thing we did added this test, because I realized while writing it that `request` does not callback with an error if it receives a 500. I had to write my own condition to get an error there.